keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1349 @@
+external-provisioner via command line parameters. There is no need to
+introduce heuristics or other, more complex ways of changing the
+behavior (like extending the `CSIDriver` API).
+
+#### Available capacity vs. maximum volume size
+
+The CSI spec up to and including the current version 1.2 just
+specifies that ["the available
+capacity"](https://github.com/container-storage-interface/spec/blob/314ac542302938640c59b6fb501c635f27015326/lib/go/csi/csi.pb.go#L2548-L2554)
+is to be returned by the driver. It is left open whether that means
+that a volume of that size can be created. This KEP uses the reported
+capacity to rule out pools which clearly have insufficient storage
+because the reported capacity is smaller than the size of a
+volume. This will work better when CSI drivers implement `GetCapacity`
+such that they consider constraints like fragmentation and report the
+size that the largest volume can have at the moment.
+
+#### Without central controller
+
+This mode of operation is expected to be used by CSI drivers that need
+to track capacity per node and only support persistent volumes with
+delayed binding that get provisioned by an external-provisioner
+instance that runs on the node where the volume gets provisioned (see
+the proposal for csi-driver-host-path in
+https://github.com/kubernetes-csi/external-provisioner/pull/367).
+
+The CSI driver has to implement the CSI controller service and its
+`GetCapacity` call. Its deployment has to add the external-provisioner
+to the daemon set and enable the per-node capacity tracking with
+`--enable-capacity=local`.
+
+The resulting `CSIStorageCapacity` objects then use a node selector for
+one node.
+
+#### With central controller
+
+For central provisioning, external-provisioner gets deployed together
+with a CSI controller service and capacity reporting gets enabled with
+`--enable-capacity=central`. In this mode, CSI drivers must report
+topology information in `NodeGetInfoResponse.accessible_topology` that
+matches the storage pool(s) that it has access to, with granularity
+that matches the most restrictive pool.
+
+For example, if the driver runs in a node with zone/region/rack
+topology and has access to one per-zone pool and one per-zone/region
+pool, then the driver should report the zone/region of the second pool
+as its topology.
+
+This is a slight deviation from the CSI spec, which defines that
+"`accessible_topology` specifies where the *node* is accessible
+from". For node-local storage there is no difference, but for
+network-attached storage there might be one.
+
+Assuming that a CSI driver meets the above requirement and enables
+this mode, external-provisioner then can identify pools as follows:
+- iterate over all `CSINode` objects and search for
+  `CSINodeDriver` information for the CSI driver,
+- compute the union of the topology segments from these
+  `CSINodeDriver` entries.
+
+For each entry in that union, one `CSIStorageCapacity` object is
```

Comment:
The name can't be determinstic. We have to use a UUID for the name, similar to how PVs are named.

Updating is still possible because external-provisioner has to know about all objects for its driver and then can match by topology and storage class name.

Reply:
PVs names are [deterministically generated](https://github.com/kubernetes-csi/external-provisioner/blob/master/pkg/controller/controller.go#L485) so that if we call CreateVolume multiple times for the same PVC, it will get called with the same name.

Searching through every object for every predicate call, or every update in external-provisioner is not ideal, especially for a local volume driver that's going to have an entry per node. Is there a way we cache the mappings (maybe keyed by node) to improve the lookups?

Reply:
> PVs names are deterministically generated so that if we call CreateVolume multiple times for the same PVC, it will get called with the same name.

That relies on the unique ID of the PVC. We have no such UID here, just a bunch of attributes. We could hash them, but then we'll have to deal with hash collisions.

Internally, both `external-provisioner` and `kube-scheduler` will have to use caching and more efficient data structures to find the information that they need.